ta/pkcs11/src/processing.c

```diff
@@ -1003,3 +1003,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+		goto out;
+	}
+
+	key = pkcs11_handle2object(key_handle, session);
+	if (!key) {
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/*
+	 * The wrapping key and key to be wrapped shouldn't be same.
+	 * PKCS#11 spec doesn't explicitly state that but logically this isn't
+	 * a use case and also acts as an attack vector, so explicitly
+	 * disallow this.
+	 */
+	if (key == wrapping_key) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	rc = set_processing_state(session, function, wrapping_key, NULL);
+	if (rc)
+		goto out;
+
+	/* Check if mechanism can be used for wrapping function */
+	rc = check_mechanism_against_processing(session, proc_params->id,
+						function,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc)
+		goto out;
+
+	/*
+	 * Check if wrapping key has CKA_WRAP set and its key type is
+	 * compatible with the mechanism passed
+	 */
+	rc = check_parent_attrs_against_processing(proc_params->id, function,
+						   wrapping_key->attributes);
+	if (rc) {
+		/*
+		 * CKR_KEY_FUNCTION_NOT_PERMITTED is not in the list of errors
+		 * specified with C_Wrap() in the specification. So
+		 * return the next most appropriate error.
+		 */
+		if (rc == PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED)
+			rc = PKCS11_CKR_WRAPPING_KEY_TYPE_INCONSISTENT;
+
+		goto out;
+	}
+
+	/* Check access of wrapping key */
+	rc = check_access_attrs_against_token(session,
+					      wrapping_key->attributes);
+	if (rc)
+		goto out;
+
+	switch (get_class(key->attributes)) {
+	case PKCS11_CKO_SECRET_KEY:
+		break;
+	case PKCS11_CKO_PRIVATE_KEY:
+		/* Key type not supported as yet */
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	default:
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
+		goto out;
+	}
+
+	/* Check if key to be wrapped is extractable */
+	if (!get_bool(key->attributes, PKCS11_CKA_EXTRACTABLE)) {
+		DMSG("Extractable property is false");
+		rc = PKCS11_CKR_KEY_UNEXTRACTABLE;
+		goto out;
+	}
+
+	if (get_bool(key->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED) &&
+	    !get_bool(wrapping_key->attributes, PKCS11_CKA_TRUSTED)) {
+		DMSG("Wrap with trusted not satisfied");
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
+		goto out;
+	}
+
+	rc = check_access_attrs_against_token(session, key->attributes);
+	if (rc)
+		goto out;
+
+	rc = get_attribute_ptr(wrapping_key->attributes,
+			       PKCS11_CKA_WRAP_TEMPLATE, (void *)&req_attrs,
+			       &size);
+	if (rc == PKCS11_CKR_OK && size != 0) {
+		if (!attributes_match_reference(key->attributes, req_attrs)) {
+			rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+			goto out;
+		}
+	}
+
+	rc = get_key_data_to_wrap(key, &key_data, &key_sz);
+	if (rc) {
+		rc = PKCS11_CKR_KEY_NOT_WRAPPABLE;
+		goto out;
+	}
+
+	if (processing_is_tee_symm(proc_params->id)) {
+		rc = init_symm_operation(session, PKCS11_FUNCTION_ENCRYPT,
+					 proc_params, wrapping_key);
+		if (rc)
+			goto out;
+
+		session->processing->mecha_type = proc_params->id;
+
+		rc  = wrap_data_by_symm_enc(session, key_data, key_sz, &out_buf,
```

Comment:
why passing `&out_ptr`, not `out_ptr` value straight?

Reply:
Will change it.